lib/parser.js

```diff
@@ -0,0 +1,99 @@
+const $RefParser = require('@apidevtools/json-schema-ref-parser');
+const { JSONPath } = require('jsonpath-plus');
+const { takeWhile, merge } = require('lodash');
+
+class ExternalComponents {
+    ref;
+    resolvedJSON;
+    constructor(ref, resolvedJSON) {
+        this.ref = ref;
+        this.resolvedJSON = resolvedJSON;
+    }
+
+    getKey(){
+        const keys = this.ref.split('/');
+        return keys[keys.length -1 ]
+    }
+
+    getValue() {
+        return this.resolvedJSON;
+    }
+}
+
+async function parse(JSONSchema) {
+    const $ref = await $RefParser.resolve(JSONSchema);
+    const refs = crawlChanelPropertiesForRefs(JSONSchema);
+    for (let ref of refs) {
+        if (isExternalReference(ref)) {
+            const componentObject = await resolveExternalRefs(JSONSchema, $ref);
+            merge(JSONSchema.components, componentObject);
+            console.log(JSONSchema.components)
```

Comment:
Never leave console log's around when it's time to merge 🙂 You can of course choose to introduce something like a logging interface others can overwrite to their liking.